### ISSUE TITLE
Memory Leak

### ISSUE THREAD

**Issue by WolfieWerewolf** [makepad/makepad.github.io#11]
Hey bro... I accidentally left the online makepad demo running overnight and when I checked the Task Manager in chrome (I'm testing on a Parallels Windows 10 VM, this morning, ) I noticed that the memory had grown substantially..

![image](https://user-images.githubusercontent.com/10762757/33515963-9a7617d6-d740-11e7-9b3f-a5ff0eeea081.png)

In case you hadn't noticed this I thought I would report it.

While I'm here I have a question... Your turtle graphics implementation.. did you write that from scratch or port it in from somewhere?  If it's borrowed do you mind sharing the original source location?  Thanks.

/W

**Comment by makepaddev** [comment]
Hey thanks for reporting, i'm rewriting everything to such an extent though i'll have to look at it again once thats done. The turtle graphics is something i wrote from scratch, ideas of course originating in the 'logo turtle'. I was thinking of ways to reduce API 'width' aka what you need to do get something, and the only real answer is 'state' also known as turtle. So i decided to figure out an entire layout and graphics api on top of it. Feel free to take any code or ideas ( i can advise just doing the ideas since the code is probably going to be hairy.)

**Comment by WolfieWerewolf** [comment]
I hear that man... I'm constantly monitoring CPU / GPU / Memory for my WebGL work that it's become a bit of an obsession... I appreciate that you are re-factoring... I've found that sharing GL resources like textures and clearing them instead of creating / destroying them is a big help..  Same with shaders, if you're layering, even in an ortho scene, and the uniforms don't change between use cases you can easily get away with re-cycling shader programs cutting back overhead related to calling into the GPU.

I'm familiar with the logo programming language which is why I find your turtle graphics implementation so interesting...  Thank you for offering for me to use it.. I may take you up on it.  I think we discussed this before but despite my years of working in JS I find JavaScript very difficult to read..  I think it's the "free-flow" nature of it.  Maybe it's my background but I find typed languages so much easier to follow... I'd really like to pull out your turtle code and play with it in my environment but I think it would be a lot of work just to identify the dependency hierarchy.  This isn't a criticism of you or JS.. It's actually a criticism of myself for not being better at reading JS.

I haven't gone through your code enough to know this but are you generating vectors from the turtle calls or textures?

Thanks

/W

**Comment by WolfieWerewolf** [comment]
I answered my own question.. going through your generated shaders now.. wow, nice work man.

**Comment by makepaddev** [comment]
Yea i'm pretty sure i just messed up not freeing some resource, i have had a long period of 'not checking if im leaking' The design isn't stupid about leaking, but i'm sure i made a bug. Also yes the reason i write JS alone is because i can almost not work with other people on large projects. I can't read their code, and i can't reason with their decisions or architecture which causes me to get stuck in a complete inability to refactor anything very quickly (or fix bugs). Whereas if i write things alone (or in tight pair-dev.. that does work) i learn about everything and i can fix bugs almost instantly, and i gather extremely useful information on what to do in my next refactor. 
Also because JS is weakly typed it really is a case of having a good memory to be able to even do it at all, even alone. The reason i do like it though is because i'm extremely fast in it, much much faster than i ever was in C++. I can undertake enormous (from a C++ point of view) projects and get somewhere, alone.
My opensource strategy is also 'please take it its free, but i'm not supporting it' Reason is because i'm not done solving the hard bits yet, and i dont wanna be slowed down. Also because i want to be free to break anything i want to improve the design, without having to weigh that decision based on the number of pissed off users, contributors or otherwise.
Someday, of course i will support makepad otherwise why attempt it. But only when i feel like its ready. On the use of code: be ware. The turtle thing is more of an interesting reusable concept than a very useful implementation for random other uses right now.
I've evolved it (rewrite from scratch) about 10 times, so its definitely useful to understand how it works since it contains a fair amount of evolutionary solutionspace.
But once you do i can really advise you to write your own because only then will you really understand what it does, and then you can also improve it. Even tell me about how you did it, so i can learn from your exploring as well.
So, bit of explaining then.
The turtle is a 'tree' of little state objects which i call turtles.
A turtle 'walks'. So if you output a rect, generally it will walk to the right the width of that rect.
This means that 'as you are drawing things', and unlike canvas, there is this thing int he back that knows the current x/y. Also it can test if the rect you are drawing would be larger than the width, and line wrap. Ok so now do this concept recursively.
So that 'rect' you draw can actually be a full box of other stuff you draw, with its own turtle. This is called 'turtle nesting', the question is how on earth do you move all this stuff to linewrap, since the draw operations 'in' that rect also just output a bit of vertexdata. That is where the magic of the layout system comes in which is 'lookbehind only'. That is i can only move things ive drawn, but i cant resize it. Since the codeflow to do the wrapping is 'inline' it can never be reexecuted magically, but i can peek behind and just 'move' the x/y values that i wrote into my vertexbuffers.
So thats the essence of it. Hope it makes sense, but i see some road ahead where i will have to explain this to people in much more detail :)

**Comment by makepaddev** [comment]
The hard bits are to fully complete a layout engine within this constraint of never resizing 'after' you called the draw method on something. For instance imagine A <B> C where B fills the space between A and C. Problem is, if you draw A B C in that order B cannot be sized correctly because it hasnt' seen you want to draw C yet, so it has no understanding of its size. So in order to do A<B>C you have to draw it like, align-right C, align left A, fill B. So thats the constraint. Sofar i'm quite happy with it

**Comment by WolfieWerewolf** [comment]
Interesting... That does help actually.. I wish I had the time to take it apart and do exactly what you said, "re-implement it and share the results" because I think together, outside of either of our environments, as a discrete UT we could probably come up with something pretty awesome around this concept alone.. I don't know about you but I have very little time for this kind of thing.. unfortunately.  I will be thinking about it though and if I do take the time to explore you can be sure I will share my findings with you... great work though, it's an impressive project especially given that you've written it by hand in JS...

**Comment by makepaddev** [comment]
And yes there is a class of UI you can't do with it. My solution is 'just hardcode the layout' or just don't do it. Life is too short to either understand or build the ultimate layout engine imho.

**Comment by WolfieWerewolf** [comment]
amen to that!

**Comment by makepaddev** [comment]
And no i don't really have time nor need to isolate this bit into a library, nothing in makepad for that matter is a library thats reusable as i prefer the system fully integrated with a single goal: make an awesome online live coding environment with unified (as far as that concept goes) CPU/GPU codespace

**Comment by WolfieWerewolf** [comment]
Yeah I get that.. I was strictly speaking from a UT perspective.... no as a re-usable implementation... I find the best way for me to understand stuff is to break it down into tiny little pieces but I get what you are saying.. Makepad is like Blender... the UI Kit is  the app, not a framework...

**Comment by makepaddev** [comment]
Sorry UT? Acronymn not expanding

**Comment by WolfieWerewolf** [comment]
Unit Test... Discrete implementation that does just one thing to prove it's functionality

**Comment by makepaddev** [comment]
Yea in that regard the layout engine is woefully undertested as well. In this last rewrite i will probably make more tests as i'm now fairly certain how this should work. Don't forget i have rewritten this from scratch probably 10 times now, i haven't yet had much use for tests except to keep throwing them away every time.

**Comment by makepaddev** [comment]
Makepad is its own (fairly large) integration test in that sense, and i do write little test html files to decomplect some things to the point i can fix bugs.

**Comment by makepaddev** [comment]
But even if i would have made extensive platform tests last cycle, i would have thrown them all away now since its getting a whole new concept of how to do out of thread APIs. I'm making the frontend of the compiler running in the user-worker with the process, outputting a serialized binary AST that gets sent to the main thread where it can either compile it into safe GC-free JS (or wasm) to talk directly to the webGL/audio apis, or into GLSL. Teeny tiny bit like a hobby LLVM architecture.

**Comment by makepaddev** [comment]
But in that cycle i hope to solve the various ways data-in and data-out can happen now on GPUS into a sortof abstract 'compute' API where you just make a compute(){ } method and have struct read and write

**Comment by WolfieWerewolf** [comment]
I noticed that you had a kernel.js and wondered if that's where you were heading :-)

**Comment by makepaddev** [comment]
I also hope to solve all of the horrible domain specific hacks that are in makepad around 2D UI handling in the renderthread

**Comment by makepaddev** [comment]
Yeah makepad essentially has a microkernel architecture

**Comment by WolfieWerewolf** [comment]
Yeah the lack of 2D api in GL is annoying.. having to depend on canvas2d to generate textures is a pain the ass...

**Comment by makepaddev** [comment]
So the idea is, you 'subclass' the RenderCommand (or in makepad its called Todo) class and provide a 'render' method. That render method can then be a little datainterpreter for whatever binary format you come up with to send to the renderer. Essentially this concept is 'drawshaders' i think, something i came up with. Its not a commandbuffer, its a program that can eat arbitrary data just like a vertexshader is for geometry and pixelshaders are for pixels.

**Comment by makepaddev** [comment]
I have this view that every time you are generalising to the point of commandbuffers essentially you made a poor mans virtual machine. And those suck, just jump to the end and make a fully programmable env that does the same.

**Comment by makepaddev** [comment]
And when wasm is mature enough i can compile them to wasm, which incidentally may make this stack extremely powerful even compared to C++ engines.

**Comment by WolfieWerewolf** [comment]
That's pretty unique and interesting.. WebCL never really took off and there's a lot of untapped potential for general purpose computing on the GPU... in particular Linear Algebra like Matrix Multiplication.. I know that's a bottleneck in my environment.. I briefly explored doing GPU math but couldn't afford the time to look deeply into it.

**Comment by WolfieWerewolf** [comment]
I agree re: WASM unfortunately it's a bit too low level at the moment to be useful.

**Comment by makepaddev** [comment]
GPU math really needs 'gigantic volumes of data' to make remote sense. Readback in webGL is damn near impossible even though we are approaching unified architectures in lots of places (just not the desktop). So that means there is a huge cliff where cpu will be faster just because of the overhead of getting data from/to the gpu

**Comment by WolfieWerewolf** [comment]
That's exactly it.. moving stuff across the threshold is expensive.. agreed.

**Comment by makepaddev** [comment]
What i'm calling out for is mostly fairly simple GPGPU things related to graphics, such as massive physics or fluid sims or otherwise interesting angles that we can explore in the interaction space of VR/AR and even webUI.

**Comment by WolfieWerewolf** [comment]
You're hardcore dude :-)

**Comment by makepaddev** [comment]
Not saying it will be 'the next way we do UI' cause that bs, but its very important to make a fun livecoding env where new things happen that are either fun, beautiful or educational to explore.

**Comment by makepaddev** [comment]
Thanks :) i think im just persistent. Been at this for 5 years or so, not nearly done.

**Comment by WolfieWerewolf** [comment]
Yeah but you've learned a LOT in those 5 years.. keep going... I've been watching you on and off for a while.. maybe a year now... (Lurking until recently)

**Comment by makepaddev** [comment]
Its an old wisdom that web fails soooo often, if you can get the compute to the data or the data to the computer, the programmer can do magic. If you cannot because you cocked up the API (i'm looking at you WEB..) suffering ensues.

**Comment by WolfieWerewolf** [comment]
Well i'm still not sold on the whole browser thing but everyone seems to want to use it so if you can't beat em join em but I'll be damned if I write anything that uses HTML again :-)

**Comment by makepaddev** [comment]
Yea cool man, always happy to get some responses! And let me know if you need more explaining, since part of my goal is to make something people understand, for me failing to explain is failure of architecture.

**Comment by WolfieWerewolf** [comment]
Well at the moment I doubt there are very many people who would understand what you've got going on in there just because you've got 5 years into it and re-written it 10 times as you said... If I had the time I'm sure I could get up to speed by which time you will have re-written it again LOL

**Comment by makepaddev** [comment]
Yea its not im a fan of browsers, at all. Its just a vehicle that exists right now with technology i can make work to reach my goal of making an awesome live coding environment. Thats why i'm also trying to make webaudio work, well enough that it will be fun to livecode synths with it. Its not a great api, but, with its limitations its very very far from being used 'at all' right now so, hopefully it wont be so intensely buggy i have to just chuck it.

**Comment by makepaddev** [comment]
But yes, untill i explicitly state i will not rewrite it tomorrow i probably will. It will just launch on makepad.io and by then i will have users to care about not pissing off, which i take seriously.

**Comment by WolfieWerewolf** [comment]
It's ironic...The people who know the most about browsers are the some ones who like browsers the least.... I appreciate what you are saying about keep users happy... and I think your approach of "don't use this or if you do you're on your own" is reasonable...

**Comment by makepaddev** [comment]
Yea i think its a nice combination of still being open, just showing everyone what you are doing is fine. If i had to do this in isolation it would be bad also for my own morale. But also not taking on weight because of that. Its funny how if you state your intentions clearly about this stuff everyone generally goes 'sure'. But if you claim things like 'this is the best thing since what not use it now its great it will make your life so much better'. and then it doesn't people are all upset :)

**Comment by WolfieWerewolf** [comment]
under promise; over deliver :-)

**Comment by WolfieWerewolf** [comment]
wolfiewerewolf@gmail.com  Feel free to ping me any time about anything.

**Comment by makepaddev** [comment]
But yeah i mean i'm super super grateful for browsers to exist though, for webGL for now webasm coming. I bitch a lot but man, holy crap we have a sandboxed env with network and cpu/gpu! Thats not even owned by one company! how great is that.

**Comment by WolfieWerewolf** [comment]
Yeah WebGL has a larger install base than Flash ever had... it is good...

**Comment by makepaddev** [comment]
And there is also a hot reloadable JIT'ing (by the best minds on the planet) dynamic language that nobody even USES for its core strength yet (live coding and prototyping)

**Comment by makepaddev** [comment]
I wouldn't even know where to begin if i had to build all of that to even get to the point of starting on makepad

**Comment by WolfieWerewolf** [comment]
All good points.. its just too bad it had to be JS.... nothing against JS its just so damn hard to read!

**Comment by makepaddev** [comment]
I have to say if typescript would be the 'main JS' but it retained all the dynamism of JS that you sometimes need i'd be totally happy with that as well. However i do need certain dynamisms to do the fairly advanced things im pulling with the compiler.

**Comment by makepaddev** [comment]
I really like the ability to runtime generate classes for instance

**Comment by WolfieWerewolf** [comment]
https://kotlinlang.org/

**Comment by makepaddev** [comment]
Yea thats pretty decent tbh

**Comment by makepaddev** [comment]
However i would be extremely surprised if i could do makepad and all im doing with kotlin

**Comment by WolfieWerewolf** [comment]
Yes you would be surprised :-)

**Comment by makepaddev** [comment]
I need a lot of introspection including turning function references into code

**Comment by makepaddev** [comment]
almost nothing that compiles does that

**Comment by WolfieWerewolf** [comment]
Well you aren't limited to "what compiles" you can call out to "native" which, in the case of compile to JS is JS....

**Comment by makepaddev** [comment]
well that kinda makes it a giant ball o mud then, i dont particularly like vm stacking to hide limitations in languages.

**Comment by makepaddev** [comment]
JS with Wasm is actually a surprisingly advanced combo. Now just a bit of optional typing, and operator overloading in JS and things would be quite nice.

**Comment by WolfieWerewolf** [comment]
The output isn't in a VM it compiles to straight ES 5.1

**Comment by makepaddev** [comment]
it transpiles to readable JS? or it LLVM barfs

**Comment by WolfieWerewolf** [comment]
its readable JS

**Comment by makepaddev** [comment]
Well then, i'd still go for making JS more readable somehow with tooling than layer another language on it tbh.

**Comment by makepaddev** [comment]
I'm not giving up yet on the code editor angle to make JS suck less for the reader.

**Comment by WolfieWerewolf** [comment]
Fair enough... to each their own :-)

**Comment by makepaddev** [comment]
Wasm will really make other languages on web bloom though, which is great. Web needs it.

**Comment by makepaddev** [comment]
To be honest i welcome all pressure on webGL, wasm, JS, only thing i dont much care for is html anymore but plenty of people who like that.

**Comment by WolfieWerewolf** [comment]
But what language will you compile to WASM?  I don't expect JS will be a WASM target, unless that's what you plan to do.

**Comment by makepaddev** [comment]
Whatever someone invents can now be fairly efficiently and compactly executed in a browser, unlike before (transpile to JS/emscripten ball-o-mud) so C#? Java? Kotlin without JS? it really just opens it up

**Comment by makepaddev** [comment]
And yes wasm is super low level, its not like JS at all, its really just the bare bones that other engines can sit on.

**Comment by makepaddev** [comment]
Oh wait i misunderstood. I'm compiling 'my typed version of JS' to Wasm.

**Comment by WolfieWerewolf** [comment]
Got it.. that's what I thought.

**Comment by makepaddev** [comment]
So makepad has shaders written in JS, i say, but thats really a lie. Its not JS its some kind of C like subset of JS that just reuses the syntax of JS but has different semantics. Its more like GLSL

**Comment by WolfieWerewolf** [comment]
Yeah GLSL generator...

**Comment by makepaddev** [comment]
Essentially i just intersected JS and GLSL and looked at what JS i could write (like objects / classes) whilst statically generating GLSL, and then applied the same constraints to generate wasm code for programmable API system.

**Comment by makepaddev** [comment]
So atleast there wouldnt be like 3 semantices but only 2.

**Comment by WolfieWerewolf** [comment]
That makes sense.. is your WASM compiler in the open source?

**Comment by makepaddev** [comment]
I'm saying im doing a wasm compiler but i really am not doing that, i'm actually generating JS right now instead of wasm because webGL is not directly accessible from wasm.

**Comment by makepaddev** [comment]
But its easier to explain if i call it a wasm compiler conceptually, and thats also where i want it to be as wasm gets more mature (gets native webGL bindings)

**Comment by makepaddev** [comment]
Everything i write will be MIT always.

**Comment by WolfieWerewolf** [comment]
not 'directly' but it can be if you pass in a JS API object when you instantiate the module.  TeaVM has a good WASM example of Box2D that does this with Canvas but the same idea would apply to WebGL.

**Comment by makepaddev** [comment]
There is no performance gain right now for me to use wasm, and considering the considerable effort i'm going for a JS implementation first.

**Comment by WolfieWerewolf** [comment]
Me too.. I'm just talking

**Comment by makepaddev** [comment]
With WASM i will need to do a lot of work, i'd rather first prove the concept of this new api tbh, which will be plenty fast for anything makepad in the next years even. But i do take the webasm constraints

**Comment by makepaddev** [comment]
I mean maybe i should target WASM and just get it over with.

**Comment by makepaddev** [comment]
* sigh *

**Comment by WolfieWerewolf** [comment]
nah I think you are right to stick with JS for now.. Let them get the kinks out of WASM and expose a higher level API.. passing ASCII chars for strings is a bit tedious..

**Comment by makepaddev** [comment]
Right? I was just going to wait it out a bit, let it stabilise, let the unreals of this world stomp their massive C++ codebase down on the poor thing and take it up in a year or maybe two

**Comment by WolfieWerewolf** [comment]
I agree.. I'm giving it a year or two as well... I just like to keep an eye on everything interesting along the way.. like Makepad!

**Comment by makepaddev** [comment]
Cool, well i'm going to keep it going. Nice talking

**Comment by WolfieWerewolf** [comment]
You too mate keep up the good work and I have no doubt we will keep in touch :-)
